influxql/iterator.gen.go.tmpl

```diff
@@ -0,0 +1,469 @@
+	for _, itr := range a {
+		itr.Close()
+	}
+	return nil
+}
+
+// new{{.Name}}Iterators converts a slice of Iterator to a slice of {{.Name}}Iterator.
+// Panic if any iterator in itrs is not a {{.Name}}Iterator.
+func new{{.Name}}Iterators(itrs []Iterator) []{{.Name}}Iterator {
+	a := make([]{{.Name}}Iterator, len(itrs))
+	for i, itr := range itrs {
+		a[i] = itr.({{.Name}}Iterator)
+	}
+	return a
+}
+
+
+// buf{{.Name}}Iterator represents a buffered {{.Name}}Iterator.
+type buf{{.Name}}Iterator struct {
+	itr {{.Name}}Iterator
+	buf *{{.Name}}Point
+}
+
+// newBuf{{.Name}}Iterator returns a buffered {{.Name}}Iterator.
+func newBuf{{.Name}}Iterator(itr {{.Name}}Iterator) *buf{{.Name}}Iterator {
+	return &buf{{.Name}}Iterator{
+		itr: itr,
+	}
+}
+
+// Close closes the underlying iterator.
+func (itr *buf{{.Name}}Iterator) Close() error { return itr.itr.Close() }
+
+// PeekTime returns the time of the next point.
+// Returns zero time if no more points available.
+func (itr *buf{{.Name}}Iterator) PeekTime() int64 {
+	v := itr.Next()
+	if v == nil {
+		return ZeroTime
+	}
+
+	t := v.Time
+	itr.unread(v)
+	return t
+}
+
+// Next returns the current buffer, if exists, or calls the underlying iterator.
+func (itr *buf{{.Name}}Iterator) Next() *{{.Name}}Point {
+	if itr.buf != nil {
+		buf := itr.buf
+		itr.buf = nil
+		return buf
+	}
+	return itr.itr.Next()
+}
+
+// Next returns the next value if it is before t.
+// If the next value is on or after t then it is moved to the buffer.
+func (itr *buf{{.Name}}Iterator) NextBefore(t int64) *{{.Name}}Point {
+	v := itr.Next()
+	if v == nil {
+		return nil
+	} else if v.Time >= t {
+		itr.unread(v)
+		return nil
+	}
+	return v
+}
+
+// unread sets v to the buffer. It is read on the next call to Next().
+func (itr *buf{{.Name}}Iterator) unread(v *{{.Name}}Point) { itr.buf = v }
+
+// buf{{.Name}}Iterators represents a list of buffered {{.Name}}Iterator.
+type buf{{.Name}}Iterators []*buf{{.Name}}Iterator
+
+// newBuf{{.Name}}Iterators returns a list of buffered {{.Name}}Iterators.
+func newBuf{{.Name}}Iterators(itrs []{{.Name}}Iterator) buf{{.Name}}Iterators {
+	a := make(buf{{.Name}}Iterators, len(itrs))
+	for i := range itrs {
+		a[i] = newBuf{{.Name}}Iterator(itrs[i])
+	}
+	return a
+}
+
+// Close closes all iterators.
+func (a buf{{.Name}}Iterators) Close() error {
+	for _, itr := range a {
+		itr.Close()
+	}
+	return nil
+}
+
+// window calculates the next window based on sorted name & tags name and time window.
+func (a buf{{.Name}}Iterators) window(opt IteratorOptions) (name string, tags Tags, endTime int64) {
+	min := ZeroTime
+	for _, itr := range a {
+		// Read next point. Ignore if no more points available.
+		p := itr.Next()
+		if p == nil {
+			continue
+		}
+
+		// Update values if point's window is lower (sorted by <name,tags,time>).
+		if min == ZeroTime || p.Name < name || (p.Name == name && p.Tags.ID() < tags.ID()) || (p.Name == name && p.Tags.Equals(&tags) && p.Time < min) {
```

Comment:
OK, so not only can an underlying iterator produce results for different groups but data from different measurements too, makes sense.